custom_envs.py

```diff
@@ -0,0 +1,111 @@
+from gym_minigrid.envs.crossing import LavaCrossingEnv
+import numpy as np
+from queue import Queue
+
+class LavaCrossingSpotRewardEnv(LavaCrossingEnv):
+  def __init__(self):
+    super().__init__()
+    self.training = True
+    self.reward_grid = None
+    self.goal_pos = (0, 0)
+    self.reset()
+
+  def _gen_reward(self):
+    # generate the reward values in the grid (use the numpy array reward_grid)
+    # first, get shortest path to goal for every square in the grid, then take
+    # the inverse of the values to get our P function (the higher it is, the
+    # closer we are to the goal
+    bfs_q = Queue()
+    processed = np.zeros_like(self.reward_grid)
+
+    # append array positions, neighbor's distance to goal to the queue, we will
+    # process a square by setting its proper distance value, adding unprocessed
+    # neighbors, modifying the # processed mask, and popping the current
+    # position from the queue
+    bfs_q.put((self.goal_pos, 0))
+    while not bfs_q.empty():
+      # parse queue item
+      pos, neighbor_dist = bfs_q.get()
+      r, c = pos
+      neighbors = [(r, c+1), (r+1, c), (r, c-1), (r-1, c)]
+
+      # if it is the goal, we don't change the reward value, otherwise, the
+      # shortest path is 1 + neighbor's distance
+      if self.reward_grid[r, c] != 2:
+        self.reward_grid[r, c] = neighbor_dist + 1
+
+      for r_n, c_n in neighbors:
+        # skip positions that are out of bounds
+        if r_n >= self.reward_grid.shape[0]: continue
```

Comment:
what happens when r_n and c_n are 0 or less than 0?

Reply:
Ah, good catch, probably the source of my bug :)